### ISSUE TITLE
maxZoomFactor should apply on top of any zoom applied by the content scale

### ISSUE THREAD

**Issue by MV-GH** [saket/telephoto#45]
Something that I have noticed is that zoomlevel seems very inconsistent. Sometimes I can barely zoom at all.

https://github.com/saket/telephoto/assets/67873169/38d91501-2883-4abc-8741-6c18a8134b36

For now I have countered this by doing 20x.

see repro: https://github.com/MV-GH/ReproChannelCrash/tree/bug/inconsisent_zoom

**Comment by saket** [comment]
I can see why this can be confusing, but the explanation is that that image is already max zoomed in to fit its bounds. This can be verified by changing the `contentScale` to `ContentScale.Inside`:

https://github.com/saket/telephoto/assets/2387680/d14b5ce7-c71e-433a-86df-0e8c3ef317a4

**Comment by MV-GH** [comment]
I see, from a technical perspective this indeed correct. But from a user perspective it does feel like one image can be zoomed in more than others. Which was the feedback i was getting. Maybe a option where it starts the zoomlevel from the size it takes up? It is not really needed for me as the users just wanted a "limitless" zoom. So I have set to a rather big number. You can close this if you want.

**Comment by saket** [comment]
Agree, I'll think more about this.

> It is not really needed for me as the users just wanted a "limitless" zoom. So I have set to a rather big number.

This unfortunately has a downside that double tapping images in your app will zoom them all the way to 20x.

**Comment by MV-GH** [comment]
I see, #32 could help with that

**Comment by FooIbar** [comment]
Yeah, we would like to have something like `maxUserZoom` instead of the current `maxZoomFactor`.

**Comment by saket** [comment]
Hi all! The next release will fix this by automatically bumping up the max zoom limit when an upscaled image is detected:

https://github.com/saket/telephoto/blob/ca0a4e2360ceff0c0146a2cb602dcb20da882255/zoomable/src/commonMain/kotlin/me/saket/telephoto/zoomable/DynamicZoomSpec.kt#L49-L57

Would anyone like to help me test this by using `0.17.0-SNAPSHOT` in your app?

```groovy
// In your settings.gradle.kts or build.gradle.kts:
repositories {
  google()
  mavenCentral()
  maven(url = "https://oss.sonatype.org/content/repositories/snapshots/")
}

// In your module's build.gradle.kts:
dependencies {
  implementation("me.saket.telephoto:zoomable-image-coil:0.17.0-SNAPSHOT")
}
```

**Comment by FooIbar** [comment]
The latest snapshot works well, but there's one thing:
https://github.com/saket/telephoto/blob/4b5dbaaaf5a83c20eb06ebf37357a61bb003ee89/zoomable/src/commonMain/kotlin/me/saket/telephoto/zoomable/DoubleClickToZoomListener.kt#L75-L79
`maxZoomFactor` here should also be shifted.

**Comment by saket** [comment]
Hmm that is going to be tricky. Do you need a custom `DoubleClickToZoomListener` anymore in your project?

**Comment by FooIbar** [comment]
I'm currently using a custom `DoubleClickToZoomListener` with `maxZoomFactor * contentTransformation.scaleMetadata.initialScale` as a workaround.
With the snapshot, if I use `ZoomSpec(maximum = ZoomLimit(factor = 2f))` and `DoubleClickToZoomListener.cycle(maxZoomFactor = 2f)`, they actually have different zoom limits if `initialScale > 1`.

**Comment by saket** [comment]
Right, but my understanding was that some apps used a custom double click listener to work around the bug where tiny images weren't zoomable. Now that that's resolved, do you still need this?

**Comment by FooIbar** [comment]
Now they are zoomable with two fingers but not double click.
Because `transformation.scale.maxScale` was not touched in 9357d58, only the max zoom factor will be bumped.

**Comment by saket** [comment]
> Now they are zoomable with two fingers but not double click.

I think there's a bit of confusion here. The images should also be zoomable with double clicks if you use the default double click listener, which doesn't read `transformation.scale.maxScale` because its `maxZoomFactor` is `null`:

https://github.com/saket/telephoto/blob/45d9fade3ecd0a4df65d0825a4ca0d7be3c25ade/zoomable/src/commonMain/kotlin/me/saket/telephoto/zoomable/DoubleClickToZoomListener.kt#L85

**Comment by FooIbar** [comment]
Ah I see what you meant.
Yes, I still need `DoubleClickToZoomListener.cycle(maxZoomFactor = 2f)`, because I want double click to be exactly 2x zoom but max zoom factor can be higher.

**Comment by saket** [comment]
Gotcha. I'll give it a shot, though I suspect I won't be able to find a way. This bugfix will likely break people's existing double click listeners since their `maxZoomFactor` won't automatically shift up.

In the meantime, I recommend calculating your `maxZoomFactor` as a percentage of `state.zoomSpec.maximum.factor`.

**Comment by FooIbar** [comment]
Well, with `ZoomSpec(maximum = ZoomLimit(factor = 2f))` and `DoubleClickToZoomListener.cycle(maxZoomFactor = 2f)`, I'd expect the image to be zoomable to 2x with both pinches and double clicks.
The old behavior is not ideal but consistent: the image can't be zoomed with either method.
The new behavior is inconsistent: it's zoomable with pinches but not double clicks.

> where tiny images weren't zoomable

BTW, it will affect (almost) all images, not just the tiny ones, with `ZoomableContentLocation.scaledInsideAndCenterAligned` and `ContentScale.FillHeight`.

**Comment by saket** [comment]
> The old behavior is not ideal but consistent: the image can't be zoomed with either method.
> The new behavior is inconsistent: it's zoomable with pinches but not double clicks.

This is unfortunately correct. If you're using a custom double click listener, you'll have to manage what `2x` zoom means yourselves. I don't think I can automatically up-shift `maxZoomFactor` because some developers might not want/need that.

I've spent two weeks trying to improve this, but I haven't been able to come up with a solution that _just_ works. If you have ideas, I'd love to hear them!

The difficulty comes from the fact that "zoom" can be defined in two different coordinate spaces: relative to the fitted size (based on the content scale) or relative to the raw image size (actual pixels). For large images that are downscaled to fit, developers usually want to mean "show me raw pixels at 2x". For tiny images that are upscaled, raw-relative zoom doesn't make sense because the image is already larger than its raw size -- so zoom must be relative to fit instead. 

This mismatch means there's no single, universal definition of "2x zoom" that works intuitively across all image sizes. `telephoto` has to balance predictability, visual fidelity, and user expectations.

**Comment by FooIbar** [comment]
I'm fine with the current implementation, but the inconsistency could be a pitfall, so it might be worth documenting?
